File: java/test/UserEntityTest.java
package voxgig.bluefinshieldconexmgmtsdk.sdktest;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import org.junit.jupiter.api.Assumptions;
import org.junit.jupiter.api.Test;

import voxgig.bluefinshieldconexmgmtsdk.core.Helpers;
import voxgig.bluefinshieldconexmgmtsdk.core.SdkEntity;
import voxgig.bluefinshieldconexmgmtsdk.core.BluefinShieldconexMgmtSDK;
import voxgig.bluefinshieldconexmgmtsdk.utility.Json;
import voxgig.bluefinshieldconexmgmtsdk.utility.struct.Struct;

@SuppressWarnings({"unchecked", "unused"})
public class UserEntityTest {

  @Test
  public void instance() {
    BluefinShieldconexMgmtSDK testsdk = BluefinShieldconexMgmtSDK.testSDK();
    SdkEntity ent = testsdk.user(null);
    assertNotNull(ent, "expected non-null user entity");
  }

  @Test
  public void basic() {
    RunnerSupport.EntityTestSetup setup = userBasicSetup(null);
    // Per-op sdk-test-control.json skip — basic test exercises a flow
    // with multiple ops; skipping any op skips the whole flow.
    String mode = setup.live ? "live" : "unit";
    for (String op : new String[] { "load" }) {
      String reason = RunnerSupport.skipReason("entityOp", "user." + op, mode);
      Assumptions.assumeTrue(reason == null,
          reason == null || "".equals(reason)
              ? "skipped via sdk-test-control.json" : reason);
    }
    // The basic flow consumes synthetic IDs from the fixture. In live mode
    // without an *_ENTID env override, those IDs hit the live API and 4xx.
    Assumptions.assumeFalse(setup.syntheticOnly,
        "live entity test uses synthetic IDs from fixture — set BLUEFINSHIELDCONEXMGMT_TEST_USER_ENTID JSON to run live");
    BluefinShieldconexMgmtSDK client = setup.client;

    // Bootstrap entity data from existing test data (no create step in flow).
    List<List<Object>> userRef01DataRaw = Struct.items(Helpers.toMapAny(
        Struct.getpath(setup.data, "existing.user")));
    Map<String, Object> userRef01Data = userRef01DataRaw.isEmpty()
        ? null : Helpers.toMapAny(userRef01DataRaw.get(0).get(1));

    // LOAD
    SdkEntity userRef01Ent = client.user(null);
    Map<String, Object> userRef01MatchDt0 = new LinkedHashMap<>();
    userRef01MatchDt0.put("id", userRef01Data.get("id"));
    Object userRef01DataDt0Loaded = userRef01Ent.load(userRef01MatchDt0, null);
    Map<String, Object> userRef01DataDt0LoadResult = Helpers.toMapAny(userRef01DataDt0Loaded);
    assertNotNull(userRef01DataDt0LoadResult, "expected load result to be a map");
    assertEquals(userRef01Data.get("id"), userRef01DataDt0LoadResult.get("id"),
        "expected load result id to match");

  }

  static RunnerSupport.EntityTestSetup userBasicSetup(Map<String, Object> extra) {
    RunnerSupport.loadEnvLocal();

    Map<String, Object> entityData;
    try {
      String entityDataSource = Files.readString(Path.of(
          "..", ".sdk", "test", "entity", "user", "UserTestData.json"));
      entityData = Helpers.toMapAny(Json.parse(entityDataSource));
    }
    catch (Exception e) {
      throw new AssertionError("failed to read user test data: " + e.getMessage(), e);
    }

    Map<String, Object> options = new LinkedHashMap<>();
    options.put("entity", entityData.get("existing"));

    BluefinShieldconexMgmtSDK client = BluefinShieldconexMgmtSDK.testSDK(options, extra);

    // Generate idmap via transform, matching TS pattern.
    List<Object> idnames = new ArrayList<>();
    idnames.add("user01");
    idnames.add("user02");
    idnames.add("user03");
    Object idmap = Struct.transform(idnames, Json.parse(
        "{\"`$PACK`\": [\"\", {"
        + "\"`$KEY`\": \"`$COPY`\","
        + "\"`$VAL`\": [\"`$FORMAT`\", \"upper\", \"`$COPY`\"]"
        + "}]}"));

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against
    // synthetic IDs from the fixture and 4xx's. Surface this so the test
    // can skip.
    String entidEnvRaw = RunnerSupport.getenv("BLUEFINSHIELDCONEXMGMT_TEST_USER_ENTID");
    boolean idmapOverridden = entidEnvRaw != null
        && entidEnvRaw.trim().startsWith("{");

    Map<String, Object> envm = new LinkedHashMap<>();
    envm.put("BLUEFINSHIELDCONEXMGMT_TEST_USER_ENTID", idmap);
    envm.put("BLUEFINSHIELDCONEXMGMT_TEST_LIVE", "FALSE");
    envm.put("BLUEFINSHIELDCONEXMGMT_TEST_EXPLAIN", "FALSE");
    envm.put("BLUEFINSHIELDCONEXMGMT_APIKEY", "NONE");
    Map<String, Object> env = RunnerSupport.envOverride(envm);

    Map<String, Object> idmapResolved = Helpers.toMapAny(env.get("BLUEFINSHIELDCONEXMGMT_TEST_USER_ENTID"));
    if (idmapResolved == null) {
      idmapResolved = Helpers.toMapAny(idmap);
    }

    boolean live = "TRUE".equals(env.get("BLUEFINSHIELDCONEXMGMT_TEST_LIVE"));
    if (live) {
      Map<String, Object> liveOpts = new LinkedHashMap<>();
      liveOpts.put("apikey", env.get("BLUEFINSHIELDCONEXMGMT_APIKEY"));
      Object mergedOpts = Struct.merge(Struct.jt(liveOpts, extra));
      client = new BluefinShieldconexMgmtSDK(Helpers.toMapAny(mergedOpts));
    }

    RunnerSupport.EntityTestSetup setup = new RunnerSupport.EntityTestSetup();
    setup.client = client;
    setup.data = entityData;
    setup.idmap = idmapResolved;
    setup.env = env;
    setup.explain = "TRUE".equals(env.get("BLUEFINSHIELDCONEXMGMT_TEST_EXPLAIN"));
    setup.live = live;
    setup.syntheticOnly = live && !idmapOverridden;
    setup.now = System.currentTimeMillis();
    return setup;
  }
}

File: kotlin/test/UserEntityTest.kt
package voxgig.bluefinshieldconexmgmtsdk.sdktest

import java.nio.file.Files
import java.nio.file.Paths

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Assumptions
import org.junit.jupiter.api.Test

import voxgig.bluefinshieldconexmgmtsdk.core.Helpers
import voxgig.bluefinshieldconexmgmtsdk.core.SdkEntity
import voxgig.bluefinshieldconexmgmtsdk.core.BluefinShieldconexMgmtSDK
import voxgig.bluefinshieldconexmgmtsdk.utility.Json
import voxgig.bluefinshieldconexmgmtsdk.utility.struct.Struct

@Suppress("UNCHECKED_CAST", "UNUSED_VARIABLE", "UNUSED_VALUE")
class UserEntityTest {

  @Test
  fun instance() {
    val testsdk = BluefinShieldconexMgmtSDK.testSDK()
    val ent = testsdk.user(null)
    assertNotNull(ent, "expected non-null user entity")
  }

  @Test
  fun basic() {
    val setup = userBasicSetup(null)
    // Per-op sdk-test-control.json skip.
    val mode = if (setup.live) "live" else "unit"
    for (op in arrayOf("load")) {
      val reason = RunnerSupport.skipReason("entityOp", "user.$op", mode)
      Assumptions.assumeTrue(
        reason == null,
        if (reason == null || "" == reason) "skipped via sdk-test-control.json" else reason,
      )
    }
    Assumptions.assumeFalse(
      setup.syntheticOnly,
      "live entity test uses synthetic IDs from fixture — set BLUEFINSHIELDCONEXMGMT_TEST_USER_ENTID JSON to run live",
    )
    val client = setup.client

    // Bootstrap entity data from existing test data (no create step in flow).
    val userRef01DataRaw = Struct.items(Helpers.toMapAny(
        Struct.getpath(setup.data, "existing.user")))
    val userRef01Data: MutableMap<String, Any?> = if (userRef01DataRaw.isEmpty())
        linkedMapOf() else (Helpers.toMapAny(userRef01DataRaw[0][1]) ?: linkedMapOf())

    // LOAD
    val userRef01Ent = client.user(null)
    val userRef01MatchDt0 = linkedMapOf<String, Any?>()
    userRef01MatchDt0["id"] = userRef01Data["id"]
    val userRef01DataDt0Loaded = userRef01Ent.load(userRef01MatchDt0, null)
    val userRef01DataDt0LoadResult = Helpers.toMapAny(userRef01DataDt0Loaded) ?: linkedMapOf()
    assertNotNull(userRef01DataDt0LoadResult, "expected load result to be a map")
    assertEquals(userRef01Data["id"], userRef01DataDt0LoadResult["id"],
        "expected load result id to match")

  }

  companion object {
    fun userBasicSetup(extra: MutableMap<String, Any?>?): RunnerSupport.EntityTestSetup {
      RunnerSupport.loadEnvLocal()

      val entityData: MutableMap<String, Any?>
      try {
        val entityDataSource = Files.readString(Paths.get(
            "..", ".sdk", "test", "entity", "user", "UserTestData.json"))
        entityData = Helpers.toMapAny(Json.parse(entityDataSource)) ?: linkedMapOf()
      } catch (e: Exception) {
        throw AssertionError("failed to read user test data: " + e.message, e)
      }

      val options = linkedMapOf<String, Any?>()
      options["entity"] = entityData["existing"]

      var client = BluefinShieldconexMgmtSDK.testSDK(options, extra)

      // Generate idmap via transform, matching TS pattern.
      val idnames = mutableListOf<Any?>()
      idnames.add("user01")
      idnames.add("user02")
      idnames.add("user03")
      val idmap = Struct.transform(idnames, Json.parse(
          "{\"`\$PACK`\": [\"\", {" +
          "\"`\$KEY`\": \"`\$COPY`\"," +
          "\"`\$VAL`\": [\"`\$FORMAT`\", \"upper\", \"`\$COPY`\"]" +
          "}]}"))

      // Detect ENTID env override before envOverride consumes it.
      val entidEnvRaw = RunnerSupport.getenv("BLUEFINSHIELDCONEXMGMT_TEST_USER_ENTID")
      val idmapOverridden = entidEnvRaw != null && entidEnvRaw.trim().startsWith("{")

      val envm = linkedMapOf<String, Any?>()
      envm["BLUEFINSHIELDCONEXMGMT_TEST_USER_ENTID"] = idmap
      envm["BLUEFINSHIELDCONEXMGMT_TEST_LIVE"] = "FALSE"
      envm["BLUEFINSHIELDCONEXMGMT_TEST_EXPLAIN"] = "FALSE"
      envm["BLUEFINSHIELDCONEXMGMT_APIKEY"] = "NONE"
      val env = RunnerSupport.envOverride(envm)

      var idmapResolved = Helpers.toMapAny(env["BLUEFINSHIELDCONEXMGMT_TEST_USER_ENTID"])
      if (idmapResolved == null) {
        idmapResolved = Helpers.toMapAny(idmap) ?: linkedMapOf()
      }

      val live = "TRUE" == env["BLUEFINSHIELDCONEXMGMT_TEST_LIVE"]
      if (live) {
        val liveOpts = linkedMapOf<String, Any?>()
        liveOpts["apikey"] = env["BLUEFINSHIELDCONEXMGMT_APIKEY"]
        val mergedOpts = Struct.merge(Struct.jt(liveOpts, extra))
        client = BluefinShieldconexMgmtSDK(Helpers.toMapAny(mergedOpts))
      }

      val setup = RunnerSupport.EntityTestSetup()
      setup.client = client
      setup.data = entityData
      setup.idmap = idmapResolved
      setup.env = env
      setup.explain = "TRUE" == env["BLUEFINSHIELDCONEXMGMT_TEST_EXPLAIN"]
      setup.live = live
      setup.syntheticOnly = live && !idmapOverridden
      setup.now = System.currentTimeMillis()
      return setup
    }
  }
}
